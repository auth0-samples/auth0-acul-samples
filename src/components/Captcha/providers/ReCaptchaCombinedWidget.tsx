import React, { useCallback, useState } from "react";
import ReCAPTCHA from "react-google-recaptcha";
import ReCAPTCHAEnterprise from "react-google-recaptcha-enterprise";

import { cn } from "@/lib/utils";

import type { CaptchaResponse, CaptchaWidgetProps } from "../index";

const RECAPTCHA_V2_PROVIDER = "recaptcha_v2";
const RECAPTCHA_ENTERPRISE_PROVIDER = "recaptcha_enterprise";

// Type alias for the component prop to be rendered
type ReCAPTCHAComponent = typeof ReCAPTCHA | typeof ReCAPTCHAEnterprise;

const RecaptchaCombinedWidget: React.FC<CaptchaWidgetProps> = ({
  config,
  onCaptchaResponse,
  className = "",
  error,
  theme,
}) => {
  const [token, setToken] = useState<string | undefined>(undefined);
  const [errorMessage, setErrorMessage] = useState<string | undefined>(error);

  const handleChange = useCallback(
    (value: string | null) => {
      const newToken = value || undefined;
      setToken(newToken);
      setErrorMessage(undefined);

      if (value) {
        const response: CaptchaResponse = {
          provider: config.provider,
          token: value,
        };
        onCaptchaResponse(response);
      } else {
        onCaptchaResponse(null);
      }
    },
    [config.provider, onCaptchaResponse]
  );

  const handleExpired = useCallback(() => {
    handleChange(null);
  }, [handleChange]);

  const handleError = useCallback(() => {
    setToken(undefined);
    onCaptchaResponse(null);
  }, [onCaptchaResponse]);

  console.log("RecaptchaCombinedWidget errorMessage:", errorMessage);

  // --- Guards ---
  const isEnterprise = config.provider === RECAPTCHA_ENTERPRISE_PROVIDER;
  const isV2 = config.provider === RECAPTCHA_V2_PROVIDER;

  if (!isEnterprise && !isV2) {
    return null;
  }

  const siteKey = config.siteKey;
  if (!siteKey) {
    return null;
  }

  // --- Helper Functions ---
  const getRecaptchaTheme = (): "light" | "dark" => {
    if (theme === "auto") {
      if (
        window.matchMedia &&
        window.matchMedia("(prefers-color-scheme: dark)").matches
      ) {
        return "dark";
      }
      return "light";
    }
    return theme || "light";
  };

  const ReCAPTCHAComponent: ReCAPTCHAComponent = isEnterprise
    ? ReCAPTCHAEnterprise
    : ReCAPTCHA;

  // --- Render ---
  const reCaptchaWidgetStyle = {
    transform: "scale(1.06)",
    transformOrigin: "0 0",
  };

  const inputName = isEnterprise
    ? "g-recaptcha-response-enterprise"
    : "g-recaptcha-response";
  const inputId = isEnterprise
    ? "hidden-recaptcha-enterprise"
    : "hidden-recaptcha-v2";

  return (
    <div className={cn("space-y-2", className)}>
      <div className="flex w-full">
        <div style={reCaptchaWidgetStyle}>
          <ReCAPTCHAComponent
            sitekey={siteKey}
            onChange={handleChange}
            onExpired={handleExpired}
            onError={handleError}
            theme={getRecaptchaTheme()}
            size="normal"
          />
        </div>
        <input type="hidden" name={inputName} id={inputId} value={token} />
      </div>
    </div>
  );
};

export default RecaptchaCombinedWidget;
