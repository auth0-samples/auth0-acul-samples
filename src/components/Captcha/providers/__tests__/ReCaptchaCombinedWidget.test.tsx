import { FormProvider, useForm } from "react-hook-form";

import { render, screen } from "@testing-library/react";

// Mock ReCAPTCHA and ReCAPTCHAEnterprise
jest.mock("react-google-recaptcha", () => ({
  __esModule: true,
  default: (props: any) => <div data-testid="recaptcha" {...props} />,
}));
jest.mock("react-google-recaptcha-enterprise", () => ({
  __esModule: true,
  default: (props: any) => (
    <div data-testid="recaptcha-enterprise" {...props} />
  ),
}));

import ReCaptchaCombinedWidget from "../ReCaptchaCombinedWidget";

const baseProps = {
  config: {
    provider: "recaptcha_v2",
    siteKey: "test-site-key",
  },
  name: "captcha",
  onCaptchaResponse: jest.fn(),
  className: "",
  theme: "light" as const,
  error: undefined,
};

function Wrapper({
  children,
  defaultValues = {},
}: {
  children: any;
  defaultValues?: any;
}) {
  const methods = useForm({ defaultValues });
  return (
    <FormProvider {...methods}>
      <form>{children({ control: methods.control })}</form>
    </FormProvider>
  );
}

describe("ReCaptchaCombinedWidget", () => {
  it("renders the ReCAPTCHA widget for v2", () => {
    render(
      <Wrapper>{() => <ReCaptchaCombinedWidget {...baseProps} />}</Wrapper>
    );
    expect(screen.getByTestId("recaptcha")).toBeInTheDocument();
  });

  it("renders the ReCAPTCHAEnterprise widget for enterprise", () => {
    const props = {
      ...baseProps,
      config: {
        provider: "recaptcha_enterprise",
        siteKey: "test-site-key",
      },
    };
    render(<Wrapper>{() => <ReCaptchaCombinedWidget {...props} />}</Wrapper>);
    expect(screen.getByTestId("recaptcha-enterprise")).toBeInTheDocument();
  });

  it("matches snapshot (default)", () => {
    const { container } = render(
      <Wrapper>{() => <ReCaptchaCombinedWidget {...baseProps} />}</Wrapper>
    );
    expect(container).toMatchSnapshot();
  });

  it("matches snapshot (with error)", () => {
    const { container } = render(
      <Wrapper>
        {() => <ReCaptchaCombinedWidget {...baseProps} error="Test error" />}
      </Wrapper>
    );
    expect(container).toMatchSnapshot();
  });
});
