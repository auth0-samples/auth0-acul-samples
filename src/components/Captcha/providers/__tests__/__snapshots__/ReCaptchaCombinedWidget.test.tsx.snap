// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ReCaptchaCombinedWidget matches snapshot (default) 1`] = `
<div>
  <form>
    <div
      class="space-y-2"
    >
      <div
        class="flex w-full"
      >
        <div
          style="transform: scale(1.06); transform-origin: 0 0;"
        >
          <div
            data-testid="recaptcha"
            sitekey="test-site-key"
            theme="light"
          />
        </div>
        <input
          id="hidden-recaptcha-v2"
          name="g-recaptcha-response"
          type="hidden"
        />
      </div>
    </div>
  </form>
</div>
`;

exports[`ReCaptchaCombinedWidget matches snapshot (with error) 1`] = `
<div>
  <form>
    <div
      class="space-y-2"
    >
      <div
        class="flex w-full"
      >
        <div
          style="transform: scale(1.06); transform-origin: 0 0;"
        >
          <div
            data-testid="recaptcha"
            sitekey="test-site-key"
            theme="light"
          />
        </div>
        <input
          id="hidden-recaptcha-v2"
          name="g-recaptcha-response"
          type="hidden"
        />
      </div>
    </div>
  </form>
</div>
`;
